src/personas/data_manager/cache.rs

```diff
@@ -56,17 +65,90 @@ impl Cache {
+                       .iter()
+                       .map(|request| DataId::Mutable(request.data_id)))
             .collect()
     }
 
+    /// Returns the next mutable data chunk we need (if any)
+    pub fn needed_mutable_chunk(&mut self) -> Option<MutableDataId> {
+        self.stop_expired_needed_mutable_chunk_request();
+
+        if self.needed_mutable_chunk_request.is_some() {
+            return None;
+        }
+
+        self.needed_mutable_chunks.iter().cloned().next()
+    }
+
+    /// Insert needed mutable data chunk.
+    pub fn insert_needed_mutable_chunk(&mut self, data_id: MutableDataId) {
+        if let Some(request) = self.needed_mutable_chunk_request.as_ref() {
+            if request.data_id == data_id {
+                return;
+            }
+        }
+
+        let _ = self.needed_mutable_chunks.insert(data_id);
+    }
+
+    /// Mark the request to retrieve the given chunk as started.
+    pub fn start_needed_mutable_chunk_request(&mut self,
+                                              data_id: MutableDataId,
+                                              msg_id: MessageId) {
+        let _ = self.needed_mutable_chunks.remove(&data_id);
+        self.needed_mutable_chunk_request = Some(ChunkRequest::new(data_id, msg_id));
+    }
+
+    /// Register successful response to a needed mutable chunk request. If we receive
+    /// responses from at least QUORUM nodes, the request is considered to be successfuly
+    /// finished.
+    pub fn handle_needed_mutable_chunk_success(&mut self,
+                                               data_id: MutableDataId,
+                                               src: XorName,
+                                               msg_id: MessageId) {
+        let done = self.needed_mutable_chunk_request
+            .as_mut()
+            .map_or(false,
+                    |request| if request.data_id == data_id && request.msg_id == msg_id {
+                        let _ = request.successes.insert(src);
+                        request.did_accumulate()
+                    } else {
+                        false
+                    });
+
+        if done {
+            self.needed_mutable_chunk_request = None;
+        }
+    }
+
+    pub fn handle_needed_mutable_chunk_failure(&mut self, src: XorName, msg_id: MessageId) {
+        let done = self.needed_mutable_chunk_request
+            .as_mut()
+            .map_or(false, |request| if request.msg_id == msg_id {
+                let _ = request.failures.insert(src);
+                !request.can_accumulate()
+            } else {
+                false
+            });
+
+        if done {
+            self.needed_mutable_chunk_request = None;
+        }
+    }
+
     /// Returns data fragments we need and want to request from other nodes.
     pub fn needed_fragments(&mut self) -> Vec<(FragmentInfo, Vec<XorName>)> {
-        self.stop_expired_requests();
+        self.stop_expired_fragment_requests();
```

Comment:
Should this method really iterate through _all_ holders and return a vector of all needed fragments? This could be a long iteration and we're probably only interested in the next bunch of requests we need to send?